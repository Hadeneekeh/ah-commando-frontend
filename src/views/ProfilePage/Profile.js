/* eslint-disable no-nested-ternary */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { Component } from 'react';
import jwtDecode from 'jwt-decode';
import PropTypes from 'prop-types';
import { Link, withRouter } from 'react-router-dom';
import Loader from 'react-loader-spinner';
import Button from '@Components/Button';
import Footer from '@Components/Footer';
import Input from '@Components/Input';
import TextArea from '@App/components/TextArea/';
import ArticleCard from '@Components/ArticleCard/';
import connectComponent from '@Lib/connect-component';
import { getProfile, editProfile, getArticles } from '@Actions/profileAction';
import { followUser } from '@Actions/followActions';
import { unFollowUser } from '@Actions/unfollowActions';
import { postImage } from '@Actions/imageAction';
import Icon from '@Components/Icon';
import RenderButton from '@Components/RenderButton';

import {
  validate,
  emailSchema,
  usernameSchema,
  bioSchema,
} from '@Utils/';
import './Profile.scss';


export class Profile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      errors: {},
      isFormValid: true,
      dialogOpen: false,
      profile: {
        username: '',
        email: '',
        image: '',
        bio: '',
        followings: [],
        followers: [],
        firstname: '',
        lastname: '',
        followerCount: '',
        followingCount: '',
      },
      article: [],
      username: '',
      userToken: '',
      isFollowing: false,
    };
    this.dialog = React.createRef();
  }

  async componentDidMount() {
    const { fetchProfile, fetchArticle } = this.props;
    // eslint-disable-next-line react/prop-types
    const { history } = this.props;
    const myUsername = history.location.pathname.split('/')[2];
    const userProfile = () => {
      const { username, email } = jwtDecode(localStorage.getItem('haven'));
      if (username === undefined) return email;
      return username;
    };
    let profileResponse;
    let articleResponse;
    if (myUsername === userProfile()) {
      profileResponse = await fetchProfile();
      articleResponse = await fetchArticle();
    } else {
      profileResponse = await fetchProfile(myUsername);
      const { id } = profileResponse.payload;
      articleResponse = await fetchArticle(id);
    }

    this.setState(prevState => ({
      ...prevState,
      profile: {
        ...prevState.profile,
        ...profileResponse.payload,
        followings: profileResponse.payload.followings,
        followers: profileResponse.payload.followers,
      },
      article: articleResponse.payload,
      userToken: userProfile(),
    }));
    const { profile: { followers }, userToken } = this.state;
    followers.forEach((fellow) => {
      if (fellow.username === userToken) {
        this.setState(prevState => ({
          ...prevState,
          isFollowing: true,
        }));
      }
    });
  }

  handleToggleEditProfileModal = (status = 'open') => {
    if (status === 'close') return this.dialog.current.close(); this.dialog.current.showModal();
  }

  renderProfileCards = (data) => (
    <div className="single-article" key={data.id}>
      <ArticleCard type="vertical" data={data} />
    </div>
  );

  handleChange = event => {
    const { name, value } = event.target;
    const errors = {};
    const [errorValue] = validate({
      [name]: value,
    }, name === 'username' ? usernameSchema : name === 'bio' ? bioSchema : emailSchema);
    errors[name] = errorValue || '';
    this.setFormValidity(errors);
    this.setState(prevState => ({
      ...prevState,
      profile: {
        ...prevState.profile,
        [name]: value,
      },
      errors,
    }));
  };

  validateForm = () => {
    const errors = {};
    const {
      profile: {
        email,
        username,
        bio,
      },
    } = this.state;
    const [emailError] = validate({ email }, emailSchema);
    const [usernameError] = validate({ username }, usernameSchema);
    const [bioError] = validate({ bio }, bioSchema);
    errors.email = emailError || '';
    errors.username = usernameError || '';
    errors.bio = bioError || '';
    return this.setFormValidity(errors);
  }

  setFormValidity = (errors) => {
    let valid = true;
    Object.values(errors).forEach((value) => {
      if (value.length > 0) {
        valid = false;
      }
    });
    this.setState({ isFormValid: valid });
    return valid;
  }


  handleImageChange = async (event) => {
    const file = event.target.files[0];
    const formData = new FormData();
    formData.append('image', file);
    const { uploadImage } = this.props;
    const response = await uploadImage(formData);
    this.setState(prevState => ({
      ...prevState,
      profile: {
        ...prevState.profile,
        image: response.payload,
      },
    }));
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const { updateProfile } = this.props;
    const {
      profile: {
        username,
        email,
        bio,
        image,
      },
    } = this.state;
    const response = await updateProfile({
      user: {
        username, email, bio, image,
      },
    });

    // set the returned data to state
    const { payload } = response;
    payload.email = email;
    const {
      username: Username,
      email: Email,
      bio: Bio,
      image: Image,
    } = payload;
    this.setState(prevState => ({
      ...prevState,
      profile: {
        ...prevState.profile,
        username: Username,
        email: Email,
        image: Image,
        bio: Bio,
      },
    }));
    return this.dialog.current.close();
  }

  handleFollowUser = async () => {
    const { follow, profile: { user: { username } } } = this.props;
    await follow(username);
    this.setState(prevState => ({
      ...prevState,
      isFollowing: true,
    }));
  };

  handleUnFollowUser = async () => {
    const { unfollow, profile: { user: { username } } } = this.props;
    await unfollow(username);
    this.setState(prevState => ({
      ...prevState,
      isFollowing: false,
    }));
  };

  render() {
    const {
      errors,
      isFormValid,
      profile:
      {
        firstname,
        username,
        lastname,
        image,
        followerCount,
        followingCount,
        bio,
        email,
      },
      article,
      isFollowing,
      userToken,
    } = this.state;
    const { image: { loading } } = this.props;
    const loader = <Loader type="BallTriangle" color="#fff" height={18} width={79} />;
    const noArticle = () => (
      <h1 className="no-article">{loading && loader}</h1>
    );
    const { profile } = this.state;
    return (
      <div data-test="profileComponent">
        <dialog ref={this.dialog}>
          <div className="container">
            <span className="close" onClick={() => this.handleToggleEditProfileModal('close')}>&times;</span>
            <h2 className="profile-heading">Edit your profile</h2>
            <form onSubmit={this.handleSubmit} encType="multipart/form-data" className="update-form">
              <input type="file" id="image" name="file" className="profile-img" onChange={this.handleImageChange} />
              <label htmlFor="image" className="form-label">
                <Icon name="camera" />
                <img src={image} alt="" className="form-image" />
              </label>
              <Input
                style={{
                  width: '100%',
                }}
                name="email"
                value={email}
                type="email"
                handleChange={this.handleChange}
                label="Email"
                error={errors.email}
              />

              <Input
                style={{
                  width: '100%',
                }}
                name="username"
                value={username}
                type="text"
                handleChange={this.handleChange}
                label="Username"
                error={errors.username}
              />

              <TextArea
                style={{
                  width: '101%',
                }}
                name="bio"
                value={bio || ''}
                type="text"
                label="Bio"
                error={errors.email}
                handleChange={this.handleChange}
              />

              <Button
                className="btn"
                label={loading ? null : 'Update Profile'}
                handleClick={this.handleSubmit}
                disabled={loading ? true : !isFormValid}
                type="submit"
                datatest="submit-button"
                style={{
                  height: '45px',
                  width: '95%',
                  marginLeft: '5%',
                  color: '#ffc700',
                  backgroundColor: '#000',
                  borderRadius: '0',
                }}
              >
                {loading && loader}
              </Button>
            </form>
          </div>
        </dialog>

        <div className="profile-container">
          <div className="sidebar">
            <ul>
              <li><Link to="/">New Article</Link></li>
              <li><Link to="/">Bookmarks</Link></li>
              <li><Link to="/">Account</Link></li>
              <li><Link to="/">Stats</Link></li>
              <li><Link to="/">Notification</Link></li>
              <li><Link to="/">Help</Link></li>
              <li><Link to="/">Log Out</Link></li>
            </ul>
          </div>
          <div className="main">
            <div className="profile-header">
              <div className="profile-details">
                <div className="name-button">
                  <h3>{`${firstname} ${lastname}`}</h3>

                  <RenderButton
                    handleToggleEditProfileModal={this.handleToggleEditProfileModal}
                    handleUnFollowUser={this.handleUnFollowUser}
                    handleFollowUser={this.handleFollowUser}
                    userToken={this.state && userToken}
                    isFollowing={this.state && isFollowing}
                    profile={this.state && profile}
                  />
                  <span style={{ marginTop: '10px' }}>
                    <p>{`@${username}`}</p>
                  </span>
                </div>
                <p>
                  {bio}
                </p>
                <p className="following-follower">
                  <span className="following">{`${followingCount} Following`}</span>
                  <span className="followers">{`${followerCount} Followers`}</span>
                </p>
              </div>
              <div className="profile-image">
                <img src={image} alt="" className="profile-image" />
              </div>
            </div>
          </div>
          <div />
          <div className="profile-links">
            <ul>
              <li className="active">
                <Link to="/">Profile</Link>
              </li>
              <li>
                <Link to="/">Likes</Link>
              </li>
              <li>
                <Link to="/">Highlighted</Link>
              </li>
              <li>
                <Link to="/">Comments</Link>
              </li>
            </ul>
          </div>
          <div />
          {article.length === 0 ? (noArticle()) : (
            <div className="profile-article" data-test="articleCard">
              {article && article.map(data => this.renderProfileCards(data))}
            </div>
          )}
        </div>
        <div className="profile-footer">
          <Footer />
        </div>
      </div>
    );
  }
}

Profile.propTypes = {
  fetchArticle: PropTypes.func.isRequired,
  fetchProfile: PropTypes.func.isRequired,
  uploadImage: PropTypes.func.isRequired,
  updateProfile: PropTypes.func.isRequired,
  image: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
  }).isRequired,
  unfollow: PropTypes.func.isRequired,
  follow: PropTypes.func.isRequired,
  profile: PropTypes.shape({
    user: PropTypes.shape({
      username: PropTypes.string,
    }),
  }).isRequired,
  history: PropTypes.shape().isRequired,
};

export default connectComponent(
  withRouter(Profile), {
    fetchProfile: (username = null) => getProfile(username),
    fetchArticle: getArticles,
    updateProfile: editProfile,
    uploadImage: postImage,
    follow: (username) => followUser(username),
    unfollow: (username) => unFollowUser(username),
  },
);
