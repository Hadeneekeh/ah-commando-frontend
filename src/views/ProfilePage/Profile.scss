.profile-container{
  padding: 20px;
  display: grid;
  grid-template-columns: 12% 75%;
  grid-column-gap: 60px;
  font-family: 'Montserrat', sans-serif;
}
.sidebar ul{
  list-style-type: none;
}
.sidebar ul li{
  list-style-type: none;
  padding-bottom: 25px;
}
.sidebar ul li a{
  text-decoration: none;
  color: #000000;
  font-size: 14px;
  font-weight: bold;
}
.profile-header{
  display: grid;
  grid-template-columns: 80% 20%;
  grid-column-gap: 30px;
  padding: 20px;
}
.profile-image{
  width: 100px;
  height: 100px;
  border-radius: 50%;
}
.profile-details h3 {
  font-weight: bold;
  padding: 0;
  margin: 0;
}
.profile-details span p {
  font-weight: bold;
  font-size: 12px;
  color: #858585;
  padding: 0;
  margin-top: -20px;
}
.profile-details p {
  font-weight: 100;
  font-size: 13px;
  padding-top: 10px;
  margin: 0;
  width: 90%;
  line-height: 23px;
}
.following{
  font-size: 12px;
  font-weight: bold;
}
.followers{
  font-size: 12px;
  margin: 20px 10px 10px;
  font-weight: bold;
}
.name-button{
  margin: 0;
  padding: 0;
  display: grid;
  grid-template-columns: 20% 30%;
}
.profile-links {
  width: 90%;
  padding-bottom: 10px;
  border-bottom: #000000 1px solid;
}

.profile-links ul{
  padding-left: 20px;
}
.profile-links ul li{
  list-style-type: none;
  float: left;
  margin-right: 30px;
}
.profile-links ul li a{
  text-decoration: none;
  color: #000000;
  font-size: 14px;
  font-weight: bold;
}
.profile-article{
  margin: 0;
  padding-top: 60px;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr
}
.single-article {
  padding: 20px;
}
.profile-footer{
  margin: 100px 0 30px 0;
}
.close {
  color: #aaaaaa;
  float: right;
  font-size: 50px;
}
.close:hover,
.close:focus {
  color: #FFC700;
  text-decoration: none;
  cursor: pointer;
}
.profile-img{
  position: relative;
  left: 25%;
  transform: translateY(-280px);
  width: 150px;
  height: 150px;
  border-radius: 50%;
  background: transparent;
  padding-top: 65px;
  padding-left: 55px;
  color: white;
}
.form-image{
  position: relative;
  left: 45%;
  opacity: .5;
  transform: translateY(-220px);
  width: 130px;
  height: 130px;
  border-radius: 50%;
  color: white;
}
.update-form{
  margin-top: 30px;
  margin-left: 20%;
}
.profile-heading{
  text-align: center;
  margin: 0;
  padding: 0;
  font-size: 15px;
}
