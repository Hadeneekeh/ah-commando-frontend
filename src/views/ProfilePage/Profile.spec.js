import { shallow } from 'enzyme';
import React from 'react';
import { findByTestAttribute } from '@Utils/';
import { Profile } from './Profile';

const state = {
  errors: {},
  isFormValid: true,
  dialogOpen: false,
  profile: {
    username: 'thorsgard',
    email: 'calix@gmail.com',
    image: 'calix.img',
    bio: 'chiboy bio',
    firstname: 'nono',
    lastname: 'calix',
    followerCount: 4,
    followingCount: 45,
    chi: 'gty',
  },
  article: [],
};

const props = {
  fetchProfile: jest.fn(() => Promise.resolve({
    payload: {
      id: 1,
      username: 'test',
      bio: 'this is my bio',
      image: 'testimage.png',
      email: 'email.gmail.com',
    },
  })),
  handleToggleEditProfileModal: jest.fn(),
  handleChange: jest.fn(),
  validateForm: jest.fn(),
  handleImageChange: jest.fn(),
  handleSubmit: jest.fn(),
  updateProfile: jest.fn(() => Promise.resolve({
    payload: {
      username: 'u', email: 'e', bio: 'b', image: 'i',
    },
  })),
  fetchArticle: jest.fn(() => Promise.resolve({
    id: 60,
    title: 'chinosons',
    articleBody: 'jghsuhuhdhdhdhd jdhjdhjdgfhjfgfj',
    description: 'chinonso djdhjijhfifjfocvivc',
    tagList: 'hiufh djhdidjhd',
    uuid: '1f1106f0',
    slug: 'chinosons-1f1106f0',
    readTime: 0,
    favorited: false,
    favoriteCounts: 0,
    image: 'http://res.cloudinary.com/dutbqk0ux/image/upload/v1564500447/gdiitdzy9l0usls1keom.png',
    authorId: 29,
    likesCount: 0,
    dislikesCount: 0,
    readCount: 0,
    createdAt: '2019-09-04T10:38:23.968Z',
    updatedAt: '2019-09-04T10:38:23.968Z',
  })),
  uploadImage: jest.fn(() => { }),
  renderProfileCards: jest.fn(() => { }),
  image: {
    loading: false,
  },
  follow: jest.fn(),
  unollow: jest.fn(),
};

describe('Profile component test', () => {
  let wrapper, instance;
  describe('have props', () => {
    beforeEach(() => {
      const setUp = (propss) => shallow(<Profile {...propss} />);
      wrapper = setUp(props);
      instance = wrapper.instance();
    });
    it('Should render without error', () => {
      const component = findByTestAttribute(wrapper, 'profileComponent');
      expect(component.length).toBe(1);
    });

    it('Should render without error', () => {
      wrapper.setState({
        article: [{
          id: 60,
          title: 'chinosons',
          articleBody: 'jghsuhuhdhdhdhd jdhjdhjdgfhjfgfj',
          description: 'chinonso djdhjijhfifjfocvivc',
          tagList: 'hiufh djhdidjhd',
          uuid: '1f1106f0',
          slug: 'chinosons-1f1106f0',
          readTime: 0,
          favorited: false,
          favoriteCounts: 0,
          image: 'http://res.cloudinary.com/dutbqk0ux/image/upload/v1564500447/gdiitdzy9l0usls1keom.png',
          authorId: 29,
          likesCount: 0,
          dislikesCount: 0,
          readCount: 0,
          createdAt: '2019-09-04T10:38:23.968Z',
          updatedAt: '2019-09-04T10:38:23.968Z',
        }],
      });
      const component = findByTestAttribute(wrapper, 'articleCard');
      expect(component.length).toBe(1);
    });

    it('should call component did mount and set username on state', () => {
      instance.componentDidMount();
      setTimeout(() => {
        expect(instance.state.profile.username).toBe('test');
        expect(instance.state.profile.email).toBe('email.gmail.com');
      }, 2000);
    });

    it('set profile to state', () => {
      const event = {
        target: {
          name: 'username',
          value: 'testuser',
        },
      };
      instance.handleChange(event);
      expect(instance.state.profile.username).toBe(event.target.value);
    });

    it('set profile to state', () => {
      wrapper.setState({
        profile: {
          email: 'nongmail.com',
          username: 'Test',
        },
      });
      instance.validateForm();
      expect(instance.setFormValidity(state.errors)).toBe(true);
    });

    it('should check for errors on form submit', () => {
      wrapper.setState({
        profile: {
          email: 'non@gmail.com',
          username: 'ttest',
          bio: 'new bio',
          image: 'image.jpg',
        },
      });
      setTimeout(() => {
        expect(instance.setFormValidity(state.errors)).toBe(false);
      }, 2000);
    });

    it('should throw an error if no username is supplied', () => {
      wrapper.setState({
        profile: {
          email: 'non@gmail.com',
          username: 'ttest',
          bio: 'new bio',
          image: 'image.jpg',
        },
      });
      instance.validateForm();
      setTimeout(() => {
        expect(instance.state.errors.email).toBe('');
        expect(instance.state.errors.username).toBe('');
      }, 2000);
    });

    it('should mock image upload', () => {
      instance.handleImageChange();
      const fileContents = 'file contents';
      const file = new Blob([fileContents], { type: 'text/plain' });
      const append = jest.fn();
      const run = jest.fn((_, evtHandler) => { evtHandler(); });
      const dummyFileReader = { run, append, result: fileContents };
      window.FormData = jest.fn(() => dummyFileReader);
      jest.spyOn(wrapper, 'setState');
      wrapper.find('input[type="file"]').simulate('change', { target: { files: [file] } });
    });

    it('handle submit event', () => {
      instance.handleSubmit();
      setTimeout(() => {
        expect(instance.state.profile.username).toBe('u');
        expect(instance.state.profile.email).toBe('e');
      }, 2000);
    });

    it('shallow render with snapshot', () => {
      const erp = shallow(<Profile {...props} />);
      expect(erp).toMatchSnapshot();
    });

    it('should check that ArticleCard exists', () => {
      const erp = shallow(<Profile {...props} />);
      setTimeout(() => {
        expect(erp.find('ArticleCard').length).toBe(1);
      });
    });
  });
});
