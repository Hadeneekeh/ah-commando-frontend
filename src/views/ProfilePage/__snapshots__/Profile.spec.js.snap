// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component test have props shallow render with snapshot 1`] = `
<div
  data-test="profileComponent"
>
  <dialog>
    <div
      className="container"
    >
      <span
        className="close"
        onClick={[Function]}
      >
        Ã—
      </span>
      <h2
        className="profile-heading"
      >
        Edit your profile
      </h2>
      <form
        className="update-form"
        encType="multipart/form-data"
        onSubmit={[Function]}
      >
        <input
          className="profile-img"
          id="image"
          name="file"
          onChange={[Function]}
          type="file"
        />
        <label
          className="form-label"
          htmlFor="image"
        >
          <Icon
            handleClick={[Function]}
            name="camera"
          />
          <img
            alt=""
            className="form-image"
            src=""
          />
        </label>
        <Input
          error=""
          handleChange={[Function]}
          handleToggle={null}
          id=""
          label="Email"
          name="email"
          placeholder=""
          style={
            Object {
              "width": "100%",
            }
          }
          togglable={false}
          type="email"
          value=""
          visible={false}
        />
        <Input
          error=""
          handleChange={[Function]}
          handleToggle={null}
          id=""
          label="Username"
          name="username"
          placeholder=""
          style={
            Object {
              "width": "100%",
            }
          }
          togglable={false}
          type="text"
          value=""
          visible={false}
        />
        <TextArea
          handleChange={[Function]}
          label="Bio"
          name="bio"
          style={
            Object {
              "width": "101%",
            }
          }
          type="text"
          value=""
        />
        <Button
          className="btn"
          datatest="submit-button"
          disabled={false}
          handleClick={[Function]}
          id=""
          label="Update Profile"
          style={
            Object {
              "backgroundColor": "#000",
              "borderRadius": "0",
              "color": "#ffc700",
              "height": "45px",
              "marginLeft": "5%",
              "width": "95%",
            }
          }
          type="submit"
        />
      </form>
    </div>
  </dialog>
  <div
    className="profile-container"
  >
    <div
      className="sidebar"
    >
      <ul>
        <li>
          <Link
            to="/"
          >
            New Article
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Bookmarks
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Account
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Stats
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Notification
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Help
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Log Out
          </Link>
        </li>
      </ul>
    </div>
    <div
      className="main"
    >
      <div
        className="profile-header"
      >
        <div
          className="profile-details"
        >
          <div
            className="name-button"
          >
            <h3>
               
            </h3>
            <withRouter(renderButton)
              handleFollowUser={[Function]}
              handleToggleEditProfileModal={[Function]}
              handleUnFollowUser={[Function]}
              isFollowing={false}
              profile={
                Object {
                  "bio": "",
                  "email": "",
                  "firstname": "",
                  "followerCount": "",
                  "followers": Array [],
                  "followingCount": "",
                  "followings": Array [],
                  "image": "",
                  "lastname": "",
                  "username": "",
                }
              }
              userToken=""
            />
            <span
              style={
                Object {
                  "marginTop": "10px",
                }
              }
            >
              <p>
                @
              </p>
            </span>
          </div>
          <p />
          <p
            className="following-follower"
          >
            <span
              className="following"
            >
               Following
            </span>
            <span
              className="followers"
            >
               Followers
            </span>
          </p>
        </div>
        <div
          className="profile-image"
        >
          <img
            alt=""
            className="profile-image"
            src=""
          />
        </div>
      </div>
    </div>
    <div />
    <div
      className="profile-links"
    >
      <ul>
        <li
          className="active"
        >
          <Link
            to="/"
          >
            Profile
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Likes
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Highlighted
          </Link>
        </li>
        <li>
          <Link
            to="/"
          >
            Comments
          </Link>
        </li>
      </ul>
    </div>
    <div />
    <h1
      className="no-article"
    />
  </div>
  <div
    className="profile-footer"
  >
    <Footer />
  </div>
</div>
`;
